source/extended-json.rst

```diff
@@ -0,0 +1,618 @@
+
+META
+====
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`_.
+
+Terms
+=====
+
+*Historical extended JSON* - A string format based on the JSON standard that
+describes a BSON document. The historical extended JSON format does not describe
+a specific, standardized format, and many tools, drivers, and libraries
+implement extended JSON in conflicting ways.
+
+*Extended JSON* - A string format based on the JSON standard that describes BSON
+documents. This specification gives a formal definition to this format.
+
+Specification
+=============
+
+Parsers
+-------
+
+An extended JSON parser (hereafter just "parser") is a tool that transforms an
+extended JSON string into another representation, such as BSON or a
+language-specific data structure. A parser MUST accept all strings in the
+canonical extended JSON format. A parser MAY also accept strings that adhere to
+other formats, such as historical extended JSON formats emitted by old versions
+of mongoexport. A parser SHOULD support at least 200 levels of nesting in an
+extended JSON document but MAY set other limits on strings it can accept as
+defined in `section 9`_ of the `JSON specification`_.
+
+.. _section 9: https://tools.ietf.org/html/rfc7159#section-9
+
+A parser can generally distinguish JSON representations of BSON types by the
+presence of a key that starts with the ``$`` symbol. For example: ``{"$symbol":
+"banana"}``. However, not every key that starts with $ will indicate the element
+represents some BSON value. If such key names of the element do not exactly
+match those of a format given in the `Canonical Format Table`_, the element MUST
+be interpreted as an ordinary JSON element. If the key names match, but the
+element cannot be parsed for other reasons (e.g. ``{"$timestamp": "-4"}``), then
+the parser MUST report an error.
+
+When parsing an extended JSON document, parsers MUST NOT consider key ordering
+as having significance. For example, the document ``{"$code": "function(){}",
+"$scope": {}}`` must be considered identical to ``{"$scope": {}, "$code":
+"function(){}"}``.
+
+.. _JSON specification: https://tools.ietf.org/html/rfc7159
+
+Generators
+----------
+
+An extended JSON generator (hereafter just "generator") produces strings in
+extended JSON. A generator MUST be able to produce strings that conform strictly
+to the canonical extended JSON format. Such generators SHOULD produce such
+strings by default. If a generator does not produce strings in the canonical
+format by default, then that generator MUST be configurable to use the canonical
+format. A generator MAY be capable of exporting strings that adhere to other
+formats, such as historical extended JSON formats emitted by old versions of
+mongoexport. A generator MUST support at least 100 levels of nesting in a BSON
```

Comment:
Is it reasonable to assume that the reader limit (i.e. "SHOULD support at least 200") is not a MUST because many drivers are using other libraries for JSON decoding, while drivers are generally encoding JSON strings themselves and more capable of adhering to a MUST for writing?